iib/common/tracing.py

```diff
@@ -0,0 +1,210 @@
+    is_class=False,
+):
+    """
+    Instrument tracing for a function or class.
+
+    :param func_or_class: The function or class to be decorated.
+    :param service_name: The name of the service to be used.
+    :param span_name: The name of the span to be created.
+    :param ignoreTracing: If True, the function will not be traced.
+    :param attributes: The attributes to be added to the span.
+    :param existing_tracer: The tracer to be used.
+    :return: The decorated function or class.
+    """
+
+    def instrument_class(cls):
+        """
+        Instruments class and filters out all the methods of a class that are to be instrumented.
+
+        :param cls: The class to be decorated.
+        :return: The decorated class.
+        """
+        for name, method in cls.__dict__.items():
+            if (
+                callable(method)
+                and not method.__name__.startswith("_")
+                and not inspect.isclass(method)
+            ):
+                setattr(cls, name, instrument_tracing(method))
+        return cls
+
+    def instrument_span(func):
+        log.info(f"Instrumenting span for {span_name}")
+        tracer = trace.get_tracer(__name__)
+        context = None
+        if trace.get_current_span():
+            context = trace.get_current_span().get_span_context()
+        else:
+            context = propagator.extract(carrier={})
+
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            log.info(f"Context inside {span_name}: {context}")
+            if kwargs.get("traceparent"):
+                log.info(f"traceparent is {kwargs.get('traceparent')}")
+                carrier = {"traceparent": kwargs.get("traceparent")}
+                trace_context = propagator.extract(carrier)
+                log.info(f"Context is {trace_context}")
+            with tracer.start_as_current_span(
+                span_name or func.__name__, kind=SpanKind.SERVER
+            ) as span:
+                # span.set_attribute("function_name", func.__name__)
+                if func.__name__:  # If the function has a name
+                    log.info(f"function_name {func.__name__}")
+                    span.set_attribute("function_name", func.__name__)
+                try:
+                    result = func(*args, **kwargs)
+                except Exception as exc:
+                    span.set_status(Status(StatusCode.ERROR))
+                    span.record_exception(exc)
+                    raise
+                else:
+                    if result:
+                        log.info(f"result {result}")
+                        span.set_attribute("result_attributes", result)
+                    if args:
+                        log.info(f"arguments {args}")
+                        span.set_attribute("arguments", args)
+                    if kwargs:
+                        # Need to handle all the types of kwargs
+                        if type(kwargs) == dict:
```

Comment:
We should use `is` instead of `==`. 
Or we more python approach `isinstance()`

`if type(kwargs) is dict:` or  if isinstance(kwargs, dict):`

More info: https://blog.devgenius.io/what-is-better-type-vs-isinstance-3b948d1b81b5#082f

Reply:
Yup changed to `is`